{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0f7e0c77-0a23-4665-9e80-39e1cb7e4945",
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "What accession number do you want to search for? Eg. CEF77095.1:  CEF77095.1\n",
      "Provide your email address (required by NCBI):  nw409@exeter.ac.uk\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sequence length: 514\n",
      "Sequence:\n",
      "MDPLVASELEKLGDGVPFRARPQHVHHTWARTFSSLPELFIQPESLVEVEKVVNLARKCRRRLVTTGCGHSPSNITCTSSWLINLDNFNKILSVNQETGVVAMEGGIRLYALCAELEKHGLTMPNLGSINEQSISGAISTGTHGSSLRHGLMSENIISLKVTLANGTTVFCSKDTKTDLFRAALLSLGAVGIITEVTFQAVPSFTLRWQQSVDTDYKMLESWNGDLWTQSEFVRVWWFPYTRRAVVWNAEKTDEELRDPPQSGYDGSIGYYVYHNLLYLAQHVPRILPWVEWLVFGMQYGFRNGTTSSAVQPSGKALLMNCLYSQFVNEWAIPLHKGPEALRRLSSWINHLTPEDPDYVPHNIPFSAEGLYVHAPVEVRVSDTTLTSNVRPYLDITADDGPTLYLNATLYRPYLMDPPCHERYYEGFEWLMKDLGGRPHWAKNFETSRVEIEAFYGKNLESFRAVRSDADPQGMFVGPWHRERIMEEGEGLELEEVEIRREKNNSGGLTTFGVI\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Do you want to use this sequence for the BLAST search? (yes/no):  yes\n",
      "Provide your email address (required by NCBI):  nw409@exeter.ac.uk\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running BLAST search (this may take a few minutes)...\n",
      "\n",
      "Top 100 BLAST hits sorted by query coverage:\n",
      "Accession     Organism                                    Query Cover Identity    E-value   \n",
      "--------------------------------------------------------------------------------------------\n",
      "XP_053005593  Fusarium falciforme                         100.2%      84.5%       0.00e+00  \n",
      "XP_052916250  Fusarium keratoplasticum                    100.2%      83.7%       0.00e+00  \n",
      "XP_046136113  Fusarium solani                             100.2%      84.1%       0.00e+00  \n",
      "UPK90849      Fusarium solani-melongenae                  100.2%      83.7%       0.00e+00  \n",
      "KAI8682984    Fusarium sp. Ph1                            100.2%      83.9%       0.00e+00  \n",
      "KAJ4321086    Fusarium piperis                            100.2%      83.9%       0.00e+00  \n",
      "KAI8681574    Fusarium keratoplasticum                    100.2%      83.7%       0.00e+00  \n",
      "KAJ4205750    Fusarium falciforme                         100.2%      83.9%       0.00e+00  \n",
      "KAJ4208211    Fusarium solani                             100.2%      83.9%       0.00e+00  \n",
      "XP_003052640  Fusarium vanettenii 77-13-4                 100.2%      83.7%       0.00e+00  \n",
      "XP_011320044  Fusarium graminearum                        100.0%      100.0%      0.00e+00  \n",
      "WXC45292      Fusarium graminearum                        100.0%      99.8%       0.00e+00  \n",
      "CAF3465019    Fusarium graminearum                        100.0%      99.8%       0.00e+00  \n",
      "KAF5228550    Fusarium austroamericanum                   100.0%      99.8%       0.00e+00  \n",
      "WXC59121      Fusarium graminearum                        100.0%      99.6%       0.00e+00  \n",
      "PTD03511      Fusarium culmorum                           100.0%      99.2%       0.00e+00  \n",
      "QPC69949      Fusarium pseudograminearum                  100.0%      98.2%       0.00e+00  \n",
      "XP_009263110  Fusarium pseudograminearum                  100.0%      98.1%       0.00e+00  \n",
      "RGP68064      Fusarium sporotrichioides                   100.0%      97.5%       0.00e+00  \n",
      "KPA41808      Fusarium langsethiae                        100.0%      97.1%       0.00e+00  \n",
      "XP_044708069  Fusarium poae                               100.0%      95.5%       0.00e+00  \n",
      "XEV03880      Fusarium sambucinum                         100.0%      95.7%       0.00e+00  \n",
      "XP_025584612  Fusarium venenatum                          100.0%      95.3%       0.00e+00  \n",
      "XP_046054027  Fusarium redolens                           100.0%      89.5%       0.00e+00  \n",
      "KAG7411805    Fusarium oxysporum                          100.0%      89.5%       0.00e+00  \n",
      "RKK73737      Fusarium oxysporum                          100.0%      88.9%       0.00e+00  \n",
      "EGU82645      Fusarium oxysporum                          100.0%      88.9%       0.00e+00  \n",
      "TVY62146      Fusarium oxysporum f. sp. cubense           100.0%      88.7%       0.00e+00  \n",
      "EXM34722      Fusarium oxysporum f. sp. vasinfectum       100.0%      88.7%       0.00e+00  \n",
      "XP_031071256  Fusarium oxysporum f. sp. cubense           100.0%      88.9%       0.00e+00  \n",
      "ENH61953      Fusarium oxysporum f. sp. cubense race 1    100.0%      88.7%       0.00e+00  \n",
      "KAF4947540    Fusarium gaditjirri                         100.0%      89.1%       0.00e+00  \n",
      "EWY88313      Fusarium oxysporum NRRL 32931               100.0%      88.7%       0.00e+00  \n",
      "KAJ4108144    Fusarium oxysporum f. sp. vasinfectum       100.0%      88.7%       0.00e+00  \n",
      "KAJ4066167    Fusarium oxysporum                          100.0%      88.5%       0.00e+00  \n",
      "KAF5266425    Fusarium oxysporum f. sp. cepae             100.0%      88.5%       0.00e+00  \n",
      "EXL44898      Fusarium oxysporum                          100.0%      88.7%       0.00e+00  \n",
      "XP_031037231  Fusarium oxysporum Fo47                     100.0%      88.7%       0.00e+00  \n",
      "KAK2696854    Fusarium oxysporum                          100.0%      88.7%       0.00e+00  \n",
      "KAH7227781    Fusarium oxysporum                          100.0%      88.7%       0.00e+00  \n",
      "KAF4449310    Fusarium austroafricanum                    100.0%      88.7%       0.00e+00  \n",
      "XP_018236955  Fusarium oxysporum f. sp. lycopersici 4287  100.0%      88.7%       0.00e+00  \n",
      "EWZ87724      Fusarium oxysporum f. sp. narcissi          100.0%      88.7%       0.00e+00  \n",
      "RKL11195      Fusarium oxysporum                          100.0%      88.7%       0.00e+00  \n",
      "EXK99108      Fusarium oxysporum f. sp. raphani           100.0%      88.5%       0.00e+00  \n",
      "KAJ4048772    Fusarium oxysporum                          100.0%      88.5%       0.00e+00  \n",
      "KAH7477755    Fusarium oxysporum f. sp. albedinis         100.0%      88.5%       0.00e+00  \n",
      "RGP59263      Fusarium longipes                           100.0%      92.8%       0.00e+00  \n",
      "KAF9765634    Fusarium sp. DS 682                         100.0%      87.5%       0.00e+00  \n",
      "KAJ4264001    Fusarium torreyae                           100.0%      90.1%       0.00e+00  \n",
      "KAF5599732    Fusarium pseudoanthophilum                  100.0%      89.7%       0.00e+00  \n",
      "PNP82502      Fusarium nygamai                            100.0%      89.7%       0.00e+00  \n",
      "KAF4424531    Fusarium acutatum                           100.0%      89.7%       0.00e+00  \n",
      "KAJ4018230    Fusarium irregulare                         100.0%      90.5%       0.00e+00  \n",
      "XP_031014809  Fusarium coffeatum                          100.0%      90.5%       0.00e+00  \n",
      "CAG7563382    Fusarium equiseti                           100.0%      90.1%       0.00e+00  \n",
      "KAF5621788    Fusarium sp. NRRL 52700                     100.0%      88.9%       0.00e+00  \n",
      "XP_045983170  Fusarium flagelliforme                      100.0%      90.5%       0.00e+00  \n",
      "KAI1069527    Fusarium sp. FIESC RH6                      100.0%      90.7%       0.00e+00  \n",
      "KAF5605889    Fusarium pseudocircinatum                   100.0%      89.1%       0.00e+00  \n",
      "KAJ4140224    Fusarium equiseti                           100.0%      90.3%       0.00e+00  \n",
      "KAF5638337    Fusarium sp. NRRL 25303                     100.0%      89.1%       0.00e+00  \n",
      "XP_036530648  Fusarium subglutinans                       100.0%      89.3%       0.00e+00  \n",
      "KAF4498048    Fusarium agapanthi                          100.0%      89.3%       0.00e+00  \n",
      "KAF5965456    Fusarium bulbicola                          100.0%      89.3%       0.00e+00  \n",
      "KAF5703121    Fusarium mundagurra                         100.0%      89.3%       0.00e+00  \n",
      "KAI1051293    Fusarium chuoi                              100.0%      89.3%       0.00e+00  \n",
      "XP_037206803  Fusarium tjaetaba                           100.0%      89.3%       0.00e+00  \n",
      "XP_023432333  Fusarium fujikuroi                          100.0%      89.1%       0.00e+00  \n",
      "XP_018745073  Fusarium verticillioides                    100.0%      89.1%       0.00e+00  \n",
      "KAF5546110    Fusarium mexicanum                          100.0%      89.3%       0.00e+00  \n",
      "XP_031082980  Fusarium proliferatum ET1                   100.0%      89.3%       0.00e+00  \n",
      "XP_044678858  Fusarium musae                              100.0%      89.1%       0.00e+00  \n",
      "KAF5985936    Fusarium coicis                             100.0%      89.3%       0.00e+00  \n",
      "KAI1015722    Fusarium chuoi                              100.0%      89.5%       0.00e+00  \n",
      "KAF4968968    Fusarium sarcochroum                        100.0%      90.1%       0.00e+00  \n",
      "RBQ86782      Fusarium verticillioides                    100.0%      88.9%       0.00e+00  \n",
      "KAF5684347    Fusarium denticulatum                       100.0%      88.7%       0.00e+00  \n",
      "KAF5715505    Fusarium globosum                           100.0%      88.9%       0.00e+00  \n",
      "KAF5673107    Fusarium circinatum                         100.0%      89.3%       0.00e+00  \n",
      "KAG4284135    Fusarium proliferatum                       100.0%      89.1%       0.00e+00  \n",
      "RBA18621      Fusarium proliferatum                       100.0%      89.1%       0.00e+00  \n",
      "KAF5567620    Fusarium napiforme                          100.0%      88.9%       0.00e+00  \n",
      "KAG4258484    Fusarium proliferatum                       100.0%      88.9%       0.00e+00  \n",
      "KAF5549272    Fusarium phyllophilum                       100.0%      88.9%       0.00e+00  \n",
      "KAG5757465    Fusarium xylarioides                        100.0%      88.5%       0.00e+00  \n",
      "KIL89656      Fusarium avenaceum                          100.0%      89.7%       0.00e+00  \n",
      "KAI1017707    Fusarium proliferatum                       100.0%      88.9%       0.00e+00  \n",
      "KAF5667941    Fusarium heterosporum                       100.0%      88.3%       0.00e+00  \n",
      "XP_041680733  Fusarium mangiferae                         100.0%      88.5%       0.00e+00  \n",
      "KAG5742752    Fusarium xylarioides                        100.0%      88.5%       0.00e+00  \n",
      "KAH7236060    Fusarium tricinctum                         100.0%      89.7%       0.00e+00  \n",
      "KAH6956692    Fusarium avenaceum                          100.0%      89.5%       0.00e+00  \n",
      "WZH48923      Fusarium acuminatum                         100.0%      89.7%       0.00e+00  \n",
      "KAG5664509    Fusarium avenaceum                          100.0%      89.7%       0.00e+00  \n",
      "KAF5250162    Fusarium anthophilum                        100.0%      89.5%       0.00e+00  \n",
      "SPJ81750      Fusarium torulosum                          100.0%      89.5%       0.00e+00  \n",
      "CAJ0548056    Fusarium sp. VM40                           100.0%      89.3%       0.00e+00  \n",
      "KAF5017260    Fusarium sp. NRRL 66182                     100.0%      89.1%       0.00e+00  \n",
      "KAF4994953    Fusarium graminum                           100.0%      87.7%       0.00e+00  \n",
      "\n",
      "Accession numbers from BLAST results:\n",
      "1. XP_053005593\n",
      "2. XP_052916250\n",
      "3. XP_046136113\n",
      "4. UPK90849\n",
      "5. KAI8682984\n",
      "6. KAJ4321086\n",
      "7. KAI8681574\n",
      "8. KAJ4205750\n",
      "9. KAJ4208211\n",
      "10. XP_003052640\n",
      "11. XP_011320044\n",
      "12. WXC45292\n",
      "13. CAF3465019\n",
      "14. KAF5228550\n",
      "15. WXC59121\n",
      "16. PTD03511\n",
      "17. QPC69949\n",
      "18. XP_009263110\n",
      "19. RGP68064\n",
      "20. KPA41808\n",
      "21. XP_044708069\n",
      "22. XEV03880\n",
      "23. XP_025584612\n",
      "24. XP_046054027\n",
      "25. KAG7411805\n",
      "26. RKK73737\n",
      "27. EGU82645\n",
      "28. TVY62146\n",
      "29. EXM34722\n",
      "30. XP_031071256\n",
      "31. ENH61953\n",
      "32. KAF4947540\n",
      "33. EWY88313\n",
      "34. KAJ4108144\n",
      "35. KAJ4066167\n",
      "36. KAF5266425\n",
      "37. EXL44898\n",
      "38. XP_031037231\n",
      "39. KAK2696854\n",
      "40. KAH7227781\n",
      "41. KAF4449310\n",
      "42. XP_018236955\n",
      "43. EWZ87724\n",
      "44. RKL11195\n",
      "45. EXK99108\n",
      "46. KAJ4048772\n",
      "47. KAH7477755\n",
      "48. RGP59263\n",
      "49. KAF9765634\n",
      "50. KAJ4264001\n",
      "51. KAF5599732\n",
      "52. PNP82502\n",
      "53. KAF4424531\n",
      "54. KAJ4018230\n",
      "55. XP_031014809\n",
      "56. CAG7563382\n",
      "57. KAF5621788\n",
      "58. XP_045983170\n",
      "59. KAI1069527\n",
      "60. KAF5605889\n",
      "61. KAJ4140224\n",
      "62. KAF5638337\n",
      "63. XP_036530648\n",
      "64. KAF4498048\n",
      "65. KAF5965456\n",
      "66. KAF5703121\n",
      "67. KAI1051293\n",
      "68. XP_037206803\n",
      "69. XP_023432333\n",
      "70. XP_018745073\n",
      "71. KAF5546110\n",
      "72. XP_031082980\n",
      "73. XP_044678858\n",
      "74. KAF5985936\n",
      "75. KAI1015722\n",
      "76. KAF4968968\n",
      "77. RBQ86782\n",
      "78. KAF5684347\n",
      "79. KAF5715505\n",
      "80. KAF5673107\n",
      "81. KAG4284135\n",
      "82. RBA18621\n",
      "83. KAF5567620\n",
      "84. KAG4258484\n",
      "85. KAF5549272\n",
      "86. KAG5757465\n",
      "87. KIL89656\n",
      "88. KAI1017707\n",
      "89. KAF5667941\n",
      "90. XP_041680733\n",
      "91. KAG5742752\n",
      "92. KAH7236060\n",
      "93. KAH6956692\n",
      "94. WZH48923\n",
      "95. KAG5664509\n",
      "96. KAF5250162\n",
      "97. SPJ81750\n",
      "98. CAJ0548056\n",
      "99. KAF5017260\n",
      "100. KAF4994953\n",
      "\n",
      "Total number of accessions: 100\n"
     ]
    }
   ],
   "source": [
    "# Import necessary modules from Biopython and other libraries\n",
    "import Bio\n",
    "import Bio.Blast\n",
    "from Bio import Entrez, SeqIO  # Entrez for accessing NCBI databases, SeqIO for sequence input/output\n",
    "from Bio.Blast import NCBIWWW  # NCBIWWW for running BLAST searches online\n",
    "from Bio.Blast import NCBIXML  # NCBIXML for parsing BLAST XML results\n",
    "from io import StringIO  # StringIO for handling string input/output\n",
    "\n",
    "def get_protein_sequence(protein_id: str) -> str:\n",
    "    \"\"\"\n",
    "    Retrieve protein sequence from NCBI using protein ID (accession number).\n",
    "    Args:\n",
    "        protein_id (str): NCBI protein ID.\n",
    "    Returns:\n",
    "        str: Protein sequence.\n",
    "    \"\"\"\n",
    "    # Prompt the user to provide their email address (required by NCBI)\n",
    "    Entrez.email = input(\"Provide your email address (required by NCBI): \")\n",
    "    \n",
    "    try:\n",
    "        # Retrieve the protein sequence from the NCBI database using the provided protein_id\n",
    "        handle = Entrez.efetch(\n",
    "            db=\"protein\",  # Specifies the database to search\n",
    "            id=protein_id,  # Uses the input protein_id here\n",
    "            rettype=\"fasta\",  # Specifies the return type as FASTA\n",
    "            retmode=\"text\"  # Specifies the return mode as text\n",
    "        )\n",
    "        \n",
    "        # Read the sequence data from the handle\n",
    "        record = SeqIO.read(handle, \"fasta\")  # Parse the FASTA data into a SeqRecord\n",
    "        handle.close()  # Close the handle to free resources\n",
    "        \n",
    "        return str(record.seq)  # Return the protein sequence as a string variable\n",
    "    \n",
    "    except Exception as e:\n",
    "        # Print an error message if fetching the sequence fails\n",
    "        print(f\"Error fetching sequence: {e}\")\n",
    "        return None  # Return None if there was an error\n",
    "\n",
    "def run_blast_search(protein_sequence: str) -> list:\n",
    "    \"\"\"\n",
    "    Run BLASTp search for a given protein sequence (option for a protein sequence already retrienved \n",
    "    with the previous function) and return top 100 hits.\n",
    "    Args:\n",
    "        protein_sequence (str): Protein sequence in FASTA format.        \n",
    "    Returns:\n",
    "        Top 100 BLAST hits sorted by query coverage and a list of their accession numbers.\n",
    "    \"\"\"\n",
    "    # Prompt the user to provide their email address (required by NCBI)\n",
    "    Entrez.email = input(\"Provide your email address (required by NCBI): \")\n",
    "    \n",
    "    try:\n",
    "        # Run BLAST search\n",
    "        print(\"Running BLAST search (this may take several minutes)...\")\n",
    "        result_handle = NCBIWWW.qblast(\n",
    "            program=\"blastp\",  # Specifies the BLAST program to use\n",
    "            database=\"nr\",  # Specifies the database to search (non-redundant protein sequences)\n",
    "            sequence=protein_sequence,  # Uses the provided protein sequence\n",
    "            hitlist_size=100,  # Limits the number of hits returned to 100\n",
    "            expect=10,  # Sets the E-value threshold for reporting hits\n",
    "            matrix_name=\"BLOSUM62\"  # Specifies the scoring matrix to use\n",
    "        )\n",
    "        \n",
    "        # Parse results\n",
    "        blast_records = NCBIXML.parse(result_handle)  # Parse the BLAST XML results\n",
    "        blast_record = next(blast_records)  # Get the first BLAST record\n",
    "        \n",
    "        # Store results\n",
    "        results = []  # Initialize a list to store the results\n",
    "        accession_numbers = []  # New list for accessions\n",
    "        for alignment in blast_record.alignments:  # Iterate over each alignment in the BLAST record\n",
    "            for hsp in alignment.hsps:  # Iterate over each high-scoring pair (HSP) in the alignment\n",
    "                query_cover = (hsp.align_length / blast_record.query_length) * 100  # Calculate query coverage\n",
    "                # Extract organism from the title\n",
    "                organism = alignment.title.split('[')[-1].strip(']') if '[' in alignment.title else \"Unknown\"\n",
    "                \n",
    "                # Append the results to the list\n",
    "                results.append({\n",
    "                    'title': alignment.title,\n",
    "                    'accession': alignment.accession,\n",
    "                    'organism': organism,\n",
    "                    'query_cover': query_cover,\n",
    "                    'e_value': hsp.expect, \n",
    "                    'identity': (hsp.identities / hsp.align_length) * 100, \n",
    "                    'score': hsp.score  \n",
    "                })\n",
    "                # Add accession to the list if not already present\n",
    "                if alignment.accession not in accession_numbers:\n",
    "                    accession_numbers.append(alignment.accession)\n",
    "        \n",
    "        # Sort results by protein query coverage\n",
    "        results.sort(key=lambda x: x['query_cover'], reverse=True)  # Sort in descending order\n",
    "        return results[:100], accession_numbers[:100]  # Return the top 100 results and accession numbers\n",
    "        \n",
    "    except Exception as e:\n",
    "        # Print an error message if the BLAST search fails\n",
    "        print(f\"Error during BLAST search: {e}\")\n",
    "        return None  # Return None if there was an error\n",
    "\n",
    "# Main execution\n",
    "protein_id = input(\"What accession number do you want to search for? Eg. CEF77095.1: \")  # Prompts user fora protein accession number\n",
    "sequence = get_protein_sequence(protein_id)  # Retrieves the protein sequence using the provided accession number\n",
    "\n",
    "if sequence:  # Check if a sequence was retrieved\n",
    "    print(f\"Sequence length: {len(sequence)}\")  # Print the length of the sequence\n",
    "    print(f\"Sequence:\\n{sequence}\")  # Print the sequence\n",
    "    \n",
    "    # Ask if the user wants to use this sequence for the BLAST search\n",
    "    use_sequence = input(\"Do you want to use this sequence for the BLAST search? (yes/no): \").strip().lower()\n",
    "    \n",
    "    if use_sequence == 'yes':  # If the user wants to proceed with the BLAST search\n",
    "        results, accessions = run_blast_search(sequence)  # Run the BLAST search with the retrieved sequence\n",
    "\n",
    "        # Print results\n",
    "        if results is not None and len(results) > 0:  # Check if results are not None and not empty\n",
    "            print(\"\\nTop 100 BLAST hits sorted by query coverage:\")\n",
    "            \n",
    "            # Set fixed widths for specific columns\n",
    "            accession_width = max(len(\"Accession\"), max(len(hit['accession']) for hit in results)) + 2\n",
    "            organism_width = max(len(\"Organism\"), max(len(hit['organism']) for hit in results)) + 2\n",
    "            query_cover_width = 12  # Fixed width based on title length\n",
    "            identity_width = 12      # Fixed width based on title length\n",
    "            e_value_width = max(len(\"E-value\"), max(len(f\"{hit['e_value']:.2e}\") for hit in results)) + 2\n",
    "\n",
    "            # Print header with dynamic widths\n",
    "            header_format = f\"{{:<{accession_width}}}{{:<{organism_width}}}{{:<{query_cover_width}}}{{:<{identity_width}}}{{:<{e_value_width}}}\"\n",
    "            print(header_format.format(\"Accession\", \"Organism\", \"Query Cover\", \"Identity\", \"E-value\"))\n",
    "            print(\"-\" * (accession_width + organism_width + query_cover_width + identity_width + e_value_width))\n",
    "\n",
    "            for hit in results:\n",
    "                print(header_format.format(\n",
    "                    hit['accession'],\n",
    "                    hit['organism'],\n",
    "                    f\"{hit['query_cover']:.1f}%\",\n",
    "                    f\"{hit['identity']:.1f}%\",\n",
    "                    f\"{hit['e_value']:.2e}\"\n",
    "                ))\n",
    "                \n",
    "            # Condense the accession list into a new variable\n",
    "            condensed_accession_list = [hit['accession'] for hit in results]\n",
    "\n",
    "            # Print accession numbers in same order as BLAST results\n",
    "            print(\"\\nAccession numbers from BLAST results:\")\n",
    "            for i, hit in enumerate(results, 1):\n",
    "                print(f\"{i}. {hit['accession']}\")\n",
    "            \n",
    "            print(f\"\\nTotal number of accessions: {len(condensed_accession_list)}\")\n",
    "        else:\n",
    "            print(\"No results found or an error occurred during the BLAST search.\")\n",
    "    else:\n",
    "        print(\"BLAST search was not performed.\")\n",
    "else:\n",
    "    print(\"No sequence retrieved or an error occurred.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "745e4498-98ac-401f-bc84-29829021577b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
